xdsl/ir/core.py

```diff
@@ -708,6 +708,33 @@ def __hash__(self):
         return hash(self._op._operands)  # pyright: ignore[reportPrivateUsage]
 
 
+class OpTraits(Iterable[OpTrait]):
+    _traits: set[OpTrait] | Callable[[], set[OpTrait]]
```

Comment:
Can we make this a dataclass? It would still not allow you to access the member right?

Reply:
the more I look at traces, the more I think that dataclasses are not really pulling their weight most of the time, in terms of being quite slow.

Reply:
But that's only on core data structures, right?
Traits is not something that is remotely on the critical path?

Reply:
At least I wouldn't want to ditch dataclasses in general, though if we can see that they are bad performance wise, then yeah